firedrake/slate/slac/compiler.py

```diff
@@ -477,24 +476,51 @@ def metaphrase_slate_to_cpp(expr, temps, prec=None):
 
     elif isinstance(expr, slate.Transpose):
         tensor, = expr.operands
-        return "(%s).transpose()" % metaphrase_slate_to_cpp(tensor, temps)
+        return "(%s).transpose()" % slate_to_cpp(tensor, temps)
 
     elif isinstance(expr, slate.Inverse):
         tensor, = expr.operands
-        return "(%s).inverse()" % metaphrase_slate_to_cpp(tensor, temps)
+        return "(%s).inverse()" % slate_to_cpp(tensor, temps)
 
     elif isinstance(expr, slate.Negative):
         tensor, = expr.operands
-        result = "-%s" % metaphrase_slate_to_cpp(tensor, temps, expr.prec)
+        result = "-%s" % slate_to_cpp(tensor, temps, expr.prec)
         return parenthesize(result, expr.prec, prec)
 
     elif isinstance(expr, (slate.Add, slate.Mul)):
         op = {slate.Add: '+',
               slate.Mul: '*'}[type(expr)]
         A, B = expr.operands
-        result = "%s %s %s" % (metaphrase_slate_to_cpp(A, temps, expr.prec),
+        result = "%s %s %s" % (slate_to_cpp(A, temps, expr.prec),
                                op,
-                               metaphrase_slate_to_cpp(B, temps, expr.prec))
+                               slate_to_cpp(B, temps, expr.prec))
+
+        return parenthesize(result, expr.prec, prec)
+
+    elif isinstance(expr, slate.IndexedTensor):
+        tensor, = expr.operands
+        idx = expr._idx
+        if expr.rank == 1:
+            # Vector segment
```

Comment:
WAT?!  How does eigen not treat vectors just like matrices :P.

Reply:
Ah, I just remembered that all of these `Eigen::MatrixBase` objects (which includes vectors as well) have a `block` method. So I'm pretty sure I can squash this into one stream of code for all cases.